oblib/parser.py

```diff
@@ -0,0 +1,216 @@
+            if the_one == None:
+                raise Exception("No entrypoint found given the set of facts")
+            else:
+                return the_one
+
+    def from_JSON_string(self, json_string):
+        """ Loads the Entrypoint from a JSON string into an entrypoint """
+
+        # Convert string to JSON data
+        json_data = json.loads(json_string)
+
+        # Perform basic validation that all required parts of the document are present.
+        if "documentType" not in json_data:
+            raise Exception("JSON is missing documentType tag")
+        if "prefixes" not in json_data:
+            raise Exception("JSON is missing prefixes tag")
+        if "dtsReferences" not in json_data:
+            raise Exception("JSON is missing dtsRererences tag")
+        if "facts" not in json_data:
+            raise Exception("JSON is missing facts tag")
+
+        # Loop through facts to determine what type of endpoint this is.
+        facts = json_data["facts"]
+        fact_names = []
+        for fact in facts:
+            fact_names.append(fact["aspects"]["xbrl:concept"])
+    
+        # Create an entrypoint.
+        entrypoint = data_model.Entrypoint(self._entrypoint_name(fact_names), self._taxonomy, dev_validation_off=True)
+
+        # Loop through facts.
+        facts = json_data["facts"]
+        for fact in facts:
+
+            # Create kwargs and populate with entity.
+            kwargs = {"entity": fact["aspects"]["xbrl:entity"]}
+
+            # TODO: id is not currently support by Entrypoint.  Uncomment when it is.
+            # if "id" in fact:
+            #     kwargs["id"] = fact["id"]
+
+            if "xbrl:periodStart" in fact["aspects"] and "xbrl:periodEnd" in fact["aspects"]:
+                if fact["aspects"]["xbrl:periodStart"] == fact["aspects"]["xbrl:periodEnd"]:
+                    kwargs["instant"] = util.convert_json_datetime(fact["aspects"]["xbrl:periodStart"])
+                else:
+                    kwargs["duration"] = {}
+                    kwargs["duration"]["start"] = util.convert_json_datetime(fact["aspects"]["xbrl:periodStart"])
+                    kwargs["duration"]["end"] = util.convert_json_datetime(fact["aspects"]["xbrl:periodEnd"])
+            else:
+                kwargs["duration"] = "forever"
+
+            if "xbrl:unit" in fact["aspects"]:
+                kwargs["unit_name"] = fact["aspects"]["xbrl:unit"]
+
+            # Add axis to kwargs if item is an axis.
+            for axis_chk in fact["aspects"]:
+                if "Axis" in axis_chk:
+                    kwargs[axis_chk.split(":")[1]] = fact["aspects"][axis_chk]
+
+            entrypoint.set(fact["aspects"]["xbrl:concept"], fact["value"], **kwargs)
+
+        return entrypoint
+
+    def from_JSON(self, filename):
+        """ Imports XBRL as JSON from the given filename. """
+
+        infile = open(filename, "r")
```

Comment:
Better as 
`
with open(filename, 'r') as infile:
    s.read()
`

Reply:
Will fix in all four places this occurs.